### DISCUSSION TITLE
renovate changes MAVEN wrapperUrl from internal registry to public

### DISCUSSION THREAD

**Discussion by P1tt187** [renovatebot/renovate#21210] [Request Help] [answered]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

34.114

### If you're self-hosting Renovate, select which platform you are using.

Bitbucket Server

### If you're self-hosting Renovate, tell us what version of the platform you run.

7.21.5

### Was this something which used to work for you, and then stopped?

I never saw this working

### Describe the bug

When renovate trys to update maven wrapper it changes the value from internal registry to maven public.
I would expect, that only the version number changes.

### Relevant debug logs

example:
`wrapperUrl=https://internal.local/maven-public/org/apache/maven/wrapper/maven-wrapper/3.1.0/maven-wrapper-3.1.0.jar`
becomes
`wrapperUrl=https://repo.maven.apache.org/maven2/org/apache/maven/wrapper/maven-wrapper/3.1.1/maven-wrapper-3.1.1.jar`

expected:
`wrapperUrl=https://internal.local/maven-public/org/apache/maven/wrapper/maven-wrapper/3.1.1/maven-wrapper-3.1.1.jar`

### Have you created a minimal reproduction repository?

No reproduction repository

**Comment by rarkins** [comment]
@benkelaar @trystian1 was this something you considered?

**Comment by trystian1** [comment]
Tought we tested this, will look at it. Thanks for reporting!

**Comment by trystian1** [comment]
Can reproduce locally, looking into a fix.
Sorry for the inconvenience.

**Comment by P1tt187** [comment]
it looks like renovate invokes `mvnw wrapper:wrapper`
you need to set the enviroment variable `MVNW_REPOURL` to change the repo url

extracting the current value and set the variable should fix the problem
maybe with something like this
`wrapperUrl=(?<registry>https?:\/\/((\w|-)\.?\/?)+)\/org\/apache\/maven\/wrapper`

Hope this is helful

P.S.: Thank you for taking care 😄

**Comment by trystian1** [comment]
Thanks! Was looking into that, thanks for thinking along!

**Comment by trystian1** [comment]
I think this fixes it.
https://github.com/renovatebot/renovate/pull/20092
it changes to url for both maven as maven wrapper if that's no problem, otherwise I will look further in that part as well.
So in the PR created it only updates maven-wrapper but changes the distributionUrl as well. 
You can check this example repo to see how it updated:
https://github.com/trystian1/yetanotherdemo/pulls
https://github.com/trystian1/yetanotherdemo/pull/6

**Comment by P1tt187** [comment]
I tested the new version, the bug is **not** fully fixed
when updateing the dependency `org/apache/maven/apache-maven` it works
when updateing the dependency `org/apache/maven/wrapper/` the same bug happens again

can you please take a look at this?

**Comment by trystian1** [comment]
I'm looking into this

**Comment by trystian1** [comment]
I ran these against latest of renovate:
https://github.com/trystian1/demorepo2/pull/18/files
https://github.com/trystian1/demorepo2/pull/17

Can you maybe rename the old PR's from renovate and close them, and also delete the branch?
Then we are sure it is not opening the old ones, and creating new ones.

**Comment by P1tt187** [comment]
as you told me i deleted a pr and also the branch
renovate opend two pr
Update dependency maven to v3.8.7 -> works 
Update dependency maven-wrapper to v3.1.1 ->  bug 🐛

I'm not sure why this happens 
the demo looks like a similar configuration
is there a way to provide more debug information?

when i look into the PR i noticed this:
```javascript
function getCustomMavenWrapperRepoUrl(
  deps: PackageDependency[]
): string | null {
  const replaceString = deps.find(
    (dep) => dep.depName === 'maven-wrapper'
  )?.replaceString;
```

it looks like that `maven-wrapper` isn't in the collection 
another possibility is 
```javascript
const match = regEx(/^(.*?)\/org\/apache\/maven\/wrapper\//).exec(
    replaceString
  );
```

maybe it doesn't match
but i'm just guessing

P.S.
thanks for your work so far ;)

**Comment by P1tt187** [comment]
testet the regex with an url that was replaced
it machtes, so i assume that `maven-wrapper` is missing

**Comment by viceice** [comment]
question is: why the wrapper isn't there?

**Comment by P1tt187** [comment]
second question: why is this working on the demo repo?

**Comment by trystian1** [comment]
So `maven-wrapper` is not updated at all or ? 
is this part in your url `org/apache/maven/wrapper/maven-wrapper`

Can you test your:
wrapperUrl=https://internal.local/maven-public/org/apache/maven/wrapper/maven-wrapper/3.1.0/maven-wrapper-3.1.0.jar

against this regex: `^(?:wrapperUrl\\s*=\\s*)(?<url>\\S*-(?<version>\\d+\\.\\d+(?:\\.\\d+)?(?:-\\w+)*)(?:.jar))`
this is used to pick the wrapper url from `(^|\\/).mvn/wrapper/maven-wrapper.properties$`

If you could send me the 
`wrapperUrl` I can locally debug why it is not picked up, or redact some parts of it. 

Or you can run it locally yourself, against a dummy repo in github. 
https://github.com/renovatebot/renovate/blob/main/docs/development/local-development.md

We sure want to fix it, but bit in the dark now what is going wrong.

**Comment by P1tt187** [comment]
in my case renovate opens a pr to update `maven-wrapper`from `3.1.0` to `3.1.1`
it changes 
- `maven-wrapper.jar`
- `maven-wrapper.properties`

The PR to update `apache-maven` was accepted before, which did not change the repo url
so this is the run after accepting `apache-maven` - maybe thats important?!

the properties file is located in the projectroot at
`.mvn/wrapper/maven-wrapper.properties`

I testet the regex, it matched on regex101-  only thing i noticed: `(?:.jar)` -> `(?:\.jar)` 

I will ask if I'm allowed to send the url to avoid compliance issues.

I try to setup local development.
I have a internal test repo to try new renovate configs. 
is it possible to run the local enviroment against this configuration?

not sure if this makes any diffrents, but, i usualy add `.gitattributes` 
```gitattributes
* text=auto eol=lf

*.bat           text eol=crlf
```

**Comment by trystian1** [comment]
@P1tt187 
Did you succeed setting up a local env?
In the company I work I see the maven-wrapper updates going well with a custom repository, so very curious what is going wrong.
Let me know if I can help with something.

**Comment by P1tt187** [comment]
Hi @trystian1 , I've managed to setup renovate in a devcontainer enviroment.
I created an internal demo repository, nearly similar to your's on github
currently i was not able to run against my testconfiguration

**Comment by P1tt187** [comment]
@trystian1 i noticed that you already reproduced the bug yourself
https://github.com/trystian1/demorepo2/pull/20/files

**Comment by P1tt187** [comment, marked as answer]
@trystian1 since 12.03.2023 the issue is not reproducable

maybe it was resolved with diffrent issue, thanks for your advice

**Reply by trystian1** [reply]
Good to hear!
